Wp 573 dynamically register services

aldafu opened the pull request:
Add support for API modules that can register one or more RPC services
themself. Modules need to export a "Module" constructor which will be
instantiated by the PZP. "Module" instance are required to have a "init"
instance method. This method takes two parameters, a register function and
an unregister function which can be used by the module to register
instantiated RPC services.

This allows API implementations to decide on their own whether to register
a service or not, depending on platform or environment.

Short example for a module implementation:

``` javascript
var MyModule = function(rpcHandler, params) {
    this.rpcHandler = rpcHandler;
    this.params = params;
};

MyModule.prototype.init = function (register, unregister) {
    // instantiate my RPC service, implemented in the same file
    var service = new MyService(this.rpcHandler, this.params);
    register(service);

    // maybe at some later point in time
    unregister(service);
};

exports.Module = MyModule;
```

Jira-issue: http://jira.webinos.org/browse/WP-573

aldafu committed:
Make util work with dependencies.json

Also fix require calls for logging util.

Jira-issue: WP-573

aldafu committed:
Refactor RPC module loading to util

Code for requiring the api service modules (from in tree) was factored
out into utility code and is not longer part of the rpc code.

Jira-issue: WP-573

aldafu committed:
Self registering API modules

Add support for API modules that can register one or more RPC services
themself. Modules need to export a "Module" constructor which will be
instantiated by the PZP. "Module" instance are required to have "init"
instance method. This method takes two parameters, a register function and
an unregister function which can be used by the module to register
instantiated RPC services.

This allows API implementations to decide on their own whether to register
a service or not, depending on platform or environment.

Short example for a module implementation:

var MyModule = function(rpcHandler, params) {
	this.rpcHandler = rpcHandler;
	this.params = params;
};

MyModule.prototype.init = function (register, unregister) {
	// instantiate my RPC service, implementd elsewhere
	var service = new MyService(this.rpcHandler, this.params);
	register(service);

	// maybe at some later point in time
	unregister(service);
};

exports.Module = MyModule;

Jira-issue: WP-573

aldafu committed:
TV API updates

* fixes dependencies.json, path to root was wrong
* variable name change for used tv mgr impl

Jira-issue: WP-573

aldafu committed:
Update TV API module to use self registered services

Don't export the service, but export a module instead which if initialized
registers the TV Manager service itself.

Jira-issue: WP-573

aldafu commented:
I've put this up for review/comments/testing.

aldafu committed:
Update Get42 example API module to use self registered services

Don't export the service, but export a module instead which if initialized
registers the service itself.

Jira-issue: WP-573

habibvirji commented on webinos/core/api/tv/lib/webinos.rpc_tv.js:
```diff
@@ -23,9 +23,19 @@
+	};
+
+	TVApiModule.prototype.init = function (register, unregister) {
+		var service = new RemoteTVManager(this.rpcHandler, this.params);
```
All modules returned by find-dependency are loaded. How does it filter module TV not to load in mobile and load in TV? Is the idea that init handle that individually?

aldafu replied:
It was the very idea to load (not register) every module, so that modules can run some code on their own and decide on their own if they want to register.

habibvirji commented on webinos/core/util/lib/loadservice.js:
```diff
@@ -0,0 +1,60 @@
+ *******************************************************************************/
+
+var webinos = require("find-dependencies")(__dirname);
+var logger = console;
```
Is webinos logging not working for this module?

aldafu committed:
loadservice: clean up required deps, use logger

Use the webinos logger, remove redundant error logging.

Jira-issue: WP-573

aldafu commented:
Pushed new commit to use webinos logger.

habibvirji commented:
This pr cannot be merged. Can you create new pr

aldafu committed:
Refactor to allow loading one or more modules

loadServiceModules function loads an array of modules while the
loadServiceModule function loads one module. This change refactors both
functions to use the same code.

Jira-issue: wP-573

aldafu committed:
Fix loading module loading from PZH admin webpage

Loading modules from the "service configuration" menu entry of the PZH
admin webpage now also uses the module loader from loadservice.js

Jira-issue: WP-573

aldafu commented:
This cannot be merged yet, as I've missed some changes related to service configuration which had changes merged yesterday.

I'm going to commit some part of the work to get this done, but more is still needed.

Again, don't merge yet. I'll announce when it's ready.

aldafu committed:
Break some long code lines

To stay in the 80 char width limit.

Jira-issue: WP-573

aldafu commented:
Ready to be tested and merged.

habibvirji commented on webinos/core/pzh/lib/pzh_otherManager.js:
```diff
@@ -45,7 +46,8 @@ var Pzh_RPC = function(_parent) {
-    self.registry.loadModules(_parent.config.serviceCache, self.rpcHandler); // load specified modules
+    loadModules(_parent.config.serviceCache, self.registry, self.rpcHandler); // load specified modules
+    self.rpcHandler.setSessionId(_parent.pzh_state.sessionId);
   };
```
Line 46 too sets sessionId. It is being twice

aldafu replied:
Thanks, fixed

aldafu committed:
Remove some leftovers.

* Remove doubled setSessionId call
* Remove unused load module function.

Jira-issue: WP-573

habibvirji commented:
Apologise still not mergeable.

aldafu commented:
Okay, I've rebased everything. Should be good now.
